rfc/001-composable-pipelines/README.md

```diff
@@ -0,0 +1,393 @@
+# Composable Pipelines
+
+- **Status**: In-Review
+- **Created**: Aug 19, 2022
+- **ETA**: Sep 19, 2022
+- **Authors**:
+  - [Matthias Vallentin](https://github.com/mavam)
+- **Contributors**:
+  - [Anthony Verez](https://github.com/netantho)
+  - [Dominik Lohmann](https://github.com/dominiklohmann)
+  - [Rémi Dettai](https://github.com/rdettai)
+  - [Thomas Peiselt](https://github.com/dispanser)
+- **Discussion**: [PR #2511](https://github.com/tenzir/vast/pull/2511)
+
+## Overview
+
+This proposal ideates a new UX around VAST's pipelines. The fundamental idea is
+that *everything is a pipeline* and build a composable UX around this concept.
+
+## Problem Statement
+
+Pipelines currently are only deployable at [predefined
+points](https://vast.io/docs/use-vast/transform). For example, they can be
+triggered for specific events on import/export, or invoked by compaction.
+
+As we are bringing pipelines to the query language, we need to solve a UX
+challenge: how do we make the data flow explicit and provide a unified
+experience, especially in the presence of custom plugins?
+
+## Solution Proposal
+
+Assume for a moment everything is a pipeline operator. Then we may write:
+
+ ```bash
+vast 'from s3://aws |
+      read json |
+      summarize count(dst) group-by src |
+      write feather |
+      to /path/to/file.feather'
```

Comment:
I think the strict separation between `write` and `to` operators doesn't always work out well: what happens if the output of the early pipeline parts is a heterogeneous dataset? We'd need multiple output streams yielding separate files (depending on the actual format). What if the output should be partitioned for uptake in a different tool that supports it (spark, _cough_)? I would want to write `to <some-blob-or-path> partition-by date`?
Both of these things are impossible because `Bytes` is opaque.

Reply:
Yeah, bytes as output of `write` might be too inflexible here.

> what happens if the output of the early pipeline parts is a heterogeneous dataset?

That would be a directory, right? I would argue that `write parquet` should demux the Arrow stream into a directory, with one Parquet file each.

So maybe we need `load` and `store` operators that couple `write`/`to` and `from`/`read`? Then you'd write:

```bash
store parquet s3:///bucket
```

This way you could add a parameter to write:

```bash
store parquet --partition-by date ...
```

So the signature would be `store<Arrow, Void>`.

Reply:
maybe separating `write` and `to` is pushing the granularity to high? @dispanser your example also reminds me of Apache Iceberg which can be seen as a "format" in many regards and writes to multiple files

Reply:
I think we can always chop down operators later when the use case warrants it. Starting with `load` and `store` that natively produce/consume Arrow feels more flexible.

Reply:
happy with this solution. We can even provide `write` and `to`, exposed or not, as buildling blocks for the `store`.

Reply:
I added a commit that introduces the two. Resolving this thread, please reopen if you see the need.